README.md
# minecraft-chunk-dumper

[![NPM version](https://img.shields.io/npm/v/minecraft-chunk-dumper.svg)](http://npmjs.com/package/minecraft-chunk-dumper)
[![Build Status](https://github.com/PrismarineJS/minecraft-chunk-dumper/workflows/CI/badge.svg)](https://github.com/PrismarineJS/minecraft-chunk-dumper/actions?query=workflow%3A%22CI%22)

Dumps chunks for minecraft versions 1.7 to 1.15 



## Install

To install a `minecraftChunkDumper` command line program, run:

```bash
npm install minecraft-chunk-dumper -g
```


## Usage

### Cli

```bash
$ minecraftChunkDumper --help

Usage:
    minecraftChunkDumper [command] <minecraft-version> <options>

Example:
    minecraftChunkDumper saveChunk "1.14.4" "chunk.dump" "chunk.meta" "chunkLight.dump" "chunkLight.meta"

Commands:
    saveChunk <minecraft-version> <chunk-file> <meta-file> [<chunk-light-file> <meta-light-file>]    save a single chunk file to specified files
    saveChunks <minecraft-version> <folder> <count>           save the specified number of chunks to the given folder
    continuouslySave <minecraft-version> <folder>             continuously saves chunks to the specified folder, until the program is stopped
```

### Programmatic example

```js
const ChunkDumper = require('minecraft-chunk-dumper')

const chunkDumper = new ChunkDumper('1.14.4')

async function run () {
  await chunkDumper.start()
  chunkDumper.on('chunk', (x, z, bitMap, chunkData) => console.log('I received a chunk at ' + x + ';' + z))
  chunkDumper.on('chunk_light', ({ chunkX, chunkZ }) => console.log('I received a chunk light at ' + chunkX + ';' + chunkZ))
  await chunkDumper.saveChunks('dumps/', 100)
  await chunkDumper.stop()
}

run().then(() => console.log('All done !'))
```

### Debugging

You can enable some debugging output using DEBUG environment variable:

DEBUG="chunk-dumper" node [...]

### API

ChunkDumper is a class which can dumps chunk for a given minecraft version.

It saves 2 type of files :
* Chunk files contain the buffer of the chunk (binary format)
* Metadata files are json files of that shape : `{"x":-10,"z":-1,"groundUp":true,"bitMap":15,"biomes":[1,2...],"blockEntities":[]}`

for more recent versions (starting from 1.14), it also saves chunk light files :
* Chunk light files contain the buffer of the chunk (binary format)
* Metadata chunk light files are json files of that shape : `{"chunkX":-10,"chunkZ":-1,"skyLightMask":15,"blockLightMask":15,"emptySkyLightMask":15,"emptyBlockLightMask":15}`

You should create an instance with the version you want. Then start it.
You then have several possibilities :
* use the "chunk" event to do whatever you want with chunks
* use saveChunk to save a single chunk
* use the saveChunks to save a given number of chunks
* use the startSavingChunks to continuously save chunks, then call stopSavingChunks to stop

When you are done, you should call the stop method to finish your session.

#### ChunkDumper(version)

Build a new ChunkDumper for minecraft `version`

#### ChunkDumper.start()

Downloads and starts the server then connect a node-minecraft-protocol client to it.
Returns a promise when ready.

#### ChunkDumper.stop()

Stops the nmp client then stops the server.
Returns a promise when finished.

#### ChunkDumper.saveChunk(chunkFile, metaFile, chunkLightFile, metaChunkLightFile, metaEntityFile)

Save 1 chunk in specified `chunkFile` and `metaFile` 
Save 1 chunk light in specified `chunkLightFile` and `metaChunkLightFile` 
Save 1 block entity in specified `metaEntityFile`
Returns a promise when finished.

#### ChunkDumper.saveChunks(folder, n)

Save n chunks in specified folder
Returns a promise when finished.

#### ChunkDumper.startSavingChunks(folder)

Continuously saves all chunk and metadata file to folder.
* Chunks are named chunk_x_z.dump
* Metadata files are named chunk_x_z.meta
* Chunk light files are named chunk_light_x_z.dump
* Metadata light files are named chunk_light_x_z.meta
* Metadata tile entity files are named tile_entity_x_y_z.meta

#### ChunkDumper.stopSavingChunks()

Stops saving chunks

#### "chunk" (x, z, groundUp, bitMap, biomes, chunkData, blockEntities)

Emitted when a chunk is received

#### "chunk_light" (chunkX, chunkZ, skyLightMask, blockLightMask, emptySkyLightMask, emptyBlockLightMask, data)

Emitted when a light update is received

#### "tile_entity" (location, action, nbtData)

Emitted when tile entity data is received

### License

MIT. Copyright (c) Romain Beaumont


HISTORY.md
# History

## 1.3.0

* close the client connection before quitting the server
* Add support for 1.17
* Update for 1.18 chunk and light packets
* Dump light data for 1.18

## 1.2.0

* save biomes and groundUp (useful for 1.15.2)

## 1.1.1

* fix chunk saving

## 1.1.0

* add chunk light support for 1.14 and beyond

## 1.0.1

* put the server.jar and server in a tmp dir instead of __dirname

## 1.0.0

* all features implemented :
   * saving one chunk, n chunks or continuously
   * through a cli or an api
   * with tests


test/util.js
const path = require('path')
module.exports = {
  makeLocalPath: (...pathElems) => path.join(__dirname, ...pathElems)
}


test/cli-test.js
/* eslint-env mocha */

const version = '1.15.2'
const assert = require('assert')
const util = require('util')
const fs = require('fs').promises
const exec = util.promisify(require('child_process').exec)
const { spawn } = require('child_process')
const fsOriginal = require('fs')
const { makeLocalPath } = require('./util')

const CMD_PATH = makeLocalPath('..', 'bin', 'cmd.js')
const CMD = 'node ' + CMD_PATH

describe('chunkDumper cli', function () {
  this.timeout(120000)
  describe('help', () => {
    it('has an help command', async () => {
      const { stdout, stderr } = await exec(CMD + ' help')
      assert.strictEqual(stderr, '')
      assert(stdout.toLowerCase().includes('usage'))
    })
    it('has a --help option', async () => {
      const { stdout, stderr } = await exec(CMD + ' --help')
      assert.strictEqual(stderr, '')
      assert(stdout.toLowerCase().includes('usage'))
    })
    it('handle no command gracefully', async () => {
      const { stdout, stderr } = await exec(CMD)
      assert.strictEqual(stderr, '')
      assert(stdout.toLowerCase().includes('usage'))
    })
  })

  describe('version', () => {
    const expectedVersion = require('../package.json').version + '\n'
    it('has a version command', async () => {
      const { stdout, stderr } = await exec(CMD + ' version')
      assert.strictEqual(stderr, '')
      assert(stdout.toLowerCase().includes(expectedVersion))
    })
    it('has a --version option', async () => {
      const { stdout, stderr } = await exec(CMD + ' --version')
      assert.strictEqual(stderr, '')
      assert(stdout.toLowerCase().includes(expectedVersion))
    })
    it('has a -v option', async () => {
      const { stdout, stderr } = await exec(CMD + ' -v')
      assert.strictEqual(stderr, '')
      assert(stdout.toLowerCase().includes(expectedVersion))
    })
  })

  it('can download one chunk', async () => {
    const filesPaths = [
      makeLocalPath('chunk.dump'),
      makeLocalPath('chunk.meta'),
      makeLocalPath('chunk_light.dump'),
      makeLocalPath('chunk_light.meta')
    ]
    const cm = `${CMD} saveChunk "${version}" ${filesPaths.map(o => `"${o}"`).join(' ')}`
    console.log('running ' + cm)
    const { stdout, stderr } = await exec(cm, { env: { DEBUG: 'chunk-dumper' } })
    console.log('stdout:' + stdout)
    console.log('stderr:' + stderr)
    assert(stdout.toLowerCase().includes('successfully'), `${stdout} should contain successfully`)
    for (const file of filesPaths) {
      await fs.access(file, fsOriginal.constants.F_OK)
      await fs.unlink(file)
    }
  })

  it('can download 10 chunks', async () => {
    const { stdout } = await exec(`${CMD} saveChunks "${version}" "${makeLocalPath('chunks')}" 10`)
    console.log(`Running: ${stdout}`)
    assert(stdout.toLowerCase().includes('successfully'))

    const dirContent = await fs.readdir(makeLocalPath('chunks'))
    assert(dirContent.length >= 40, 'should have at least 40 files')
    for (const file of dirContent) {
      await fs.unlink(makeLocalPath('chunks', file))
    }
    await fs.rmdir(makeLocalPath('chunks'))
  })

  it('can continuously save chunks', async () => {
    await new Promise((resolve, reject) => {
      const child = spawn('node', [CMD_PATH, 'continuouslySave', version, makeLocalPath('chunks')])
      child.on('error', reject)

      child.stdout.on('data', (data) => {
        console.log('stdout: ' + data)
        if (data.includes('Saving chunks')) {
          setTimeout(() => child.kill('SIGINT'), 10000)
        }
      })

      child.on('close', async () => {
        const dirContent = await fs.readdir(makeLocalPath('chunks'))
        assert.notStrictEqual(dirContent.length, 0)
        for (const file of dirContent) {
          await fs.unlink(makeLocalPath('chunks', file))
        }
        await fs.rmdir(makeLocalPath('chunks'))
        resolve()
      })
    })
  })
})


test/lib-test.js
/* eslint-env mocha */

const version = '1.15.2'
const ChunkDumper = require('../index.js')
const chunkDumper = new ChunkDumper(version)
const fs = require('fs').promises
const path = require('path')
const assert = require('assert')
const fsOriginal = require('fs')
const { once } = require('events')
const end = async ({ client }) => {
  if (client.state !== 'play') await once(chunkDumper.client, 'login')
  client.end()
  await once(client, 'end')
}

describe('chunkDumper lib', function () {
  this.timeout(120000)
  before(async () => {
    await chunkDumper.start()
    await end(chunkDumper)
  })
  beforeEach(() => {
    chunkDumper.logBackIn()
  })
  afterEach(async () => {
    await end(chunkDumper)
  })
  after(async () => {
    await chunkDumper.stop()
  })
  it('can receive a chunk event', async () => {
    await new Promise(resolve => chunkDumper.on('chunk', () => resolve()))
  })

  it('can save a chunk', async () => {
    await chunkDumper.saveChunk(path.join(__dirname, 'chunk.dump'), path.join(__dirname, 'chunk.meta'),
      path.join(__dirname, 'chunk_light.dump'), path.join(__dirname, 'chunk_light.meta'))
    await fs.access(path.join(__dirname, 'chunk.dump'), fsOriginal.constants.F_OK)
    await fs.access(path.join(__dirname, 'chunk.meta'), fsOriginal.constants.F_OK)
    await fs.access(path.join(__dirname, 'chunk_light.dump'), fsOriginal.constants.F_OK)
    await fs.access(path.join(__dirname, 'chunk_light.meta'), fsOriginal.constants.F_OK)
    await fs.unlink(path.join(__dirname, 'chunk.dump'))
    await fs.unlink(path.join(__dirname, 'chunk.meta'))
    await fs.unlink(path.join(__dirname, 'chunk_light.dump'))
    await fs.unlink(path.join(__dirname, 'chunk_light.meta'))
  })

  it('can save 10 chunks', async () => {
    await chunkDumper.saveChunks(path.join(__dirname, 'chunks'), 10)
    const dirContent = await fs.readdir(path.join(__dirname, 'chunks'))
    assert(dirContent.length >= 40, 'should have at least 40 files, instead had ' + dirContent.length)
    for (const file of dirContent) {
      await fs.unlink(path.join(path.join(__dirname, 'chunks'), file))
    }
    await fs.rmdir(path.join(__dirname, 'chunks'))
  })

  it('can save chunks continuously', async () => {
    chunkDumper.startSavingChunks(path.join(__dirname, 'chunks'))
    await new Promise((resolve) => setTimeout(() => {
      chunkDumper.stopSavingChunks(path.join(__dirname, 'chunks'))
      resolve()
    }, 10000))

    const dirContent = await fs.readdir(path.join(__dirname, 'chunks'))
    assert.notStrictEqual(dirContent.length, 0)
    for (const file of dirContent) {
      await fs.unlink(path.join(path.join(__dirname, 'chunks'), file))
    }
    await fs.rmdir(path.join(__dirname, 'chunks'))
  })
})


index.js
const EventEmitter = require('events').EventEmitter
const WrapServer = require('minecraft-wrap').WrapServer
const downloadServer = require('minecraft-wrap').downloadServer
const path = require('path')
const os = require('os')
const MC_SERVER_PATH = path.join(os.tmpdir(), 'server')
const MC_SERVER_JAR = path.join(os.tmpdir(), 'server.jar')
const fs = require('fs').promises
const util = require('util')
const mc = require('minecraft-protocol')
const debug = require('debug')('chunk-dumper')

class ChunkDumper extends EventEmitter {
  constructor (version) {
    super()
    this.version = version.toString()
    this.mcData = require('minecraft-data')(this.version)
    this.withLightPackets = this.mcData.isNewerOrEqualTo('1.14') && this.mcData.isOlderThan('1.18')
  }

  async start () {
    debug('downloading server')
    await util.promisify(downloadServer)(this.version, MC_SERVER_JAR)
    this.server = new WrapServer(MC_SERVER_JAR, MC_SERVER_PATH)
    this.server.startServerAsync = util.promisify(this.server.startServer)
    this.server.stopServerAsync = util.promisify(this.server.stopServer)
    this.server.deleteServerDataAsync = util.promisify(this.server.deleteServerData)

    debug('starting server')
    this.server.on('line', (line) => {
      debug(line)
    })
    await this.server.startServerAsync({ 'server-port': 25569, 'online-mode': 'false' })
    debug('connecting client')
    this.client = mc.createClient({
      username: 'Player',
      version: this.version,
      port: 25569
    })
    this.client.on('map_chunk', ({ x, z, groundUp, bitMap, biomes, chunkData, heightmaps, skyLightMask, blockLightMask, emptySkyLightMask, emptyBlockLightMask, skyLight, blockLight, trustEdges }) => {
      this.emit('chunk', ({ x, z, groundUp, bitMap, biomes, chunkData, heightmaps, skyLightMask, blockLightMask, emptySkyLightMask, emptyBlockLightMask, skyLight, blockLight, trustEdges }))
    })
    this.client.on('update_light', ({ chunkX, chunkZ, skyLightMask, blockLightMask, emptySkyLightMask, emptyBlockLightMask, skyLight, blockLight, data, trustEdges }) => {
      this.emit('chunk_light', ({ chunkX, chunkZ, skyLightMask, blockLightMask, emptySkyLightMask, emptyBlockLightMask, skyLight, blockLight, data, trustEdges }))
    })
  }

  async stop () {
    this.client.end()
    debug('stopping server')
    await this.server.stopServerAsync()
    debug('deleting data')
    await this.server.deleteServerDataAsync()
    debug('deleting server')
    await fs.unlink(MC_SERVER_JAR)
    await fs.unlink(path.join(process.cwd(), 'versions', this.version, this.version + '.json'))
    await fs.rmdir(path.join(process.cwd(), 'versions', this.version))
    await fs.rmdir(path.join(process.cwd(), 'versions'))
  }

  async saveChunk (chunkFile, metaFile, chunkLightFile, lightMetaFile) {
    await this.saveChunks('', 1, {
      chunkFile,
      metaFile,
      chunkLightFile,
      lightMetaFile
    })
  }

  async saveChunks (folder, count, forcedFileNames = undefined) {
    try {
      await fs.mkdir(folder, { recursive: true })
    } catch (err) {

    }
    const lightsSaved = new Set()
    const chunksSaved = new Set()
    await new Promise((resolve, reject) => {
      let saveChunkLight
      const saveChunk = async d => {
        const { x, z } = d
        chunksSaved.add(`${x},${z}`)
        let finished = false

        if ((chunksSaved.size === count && !this.withLightPackets) || (([...lightsSaved].filter(x => chunksSaved.has(x))).length >= count &&
          this.withLightPackets)) {
          this.removeListener('chunk', saveChunk)
          if (this.withLightPackets) {
            this.removeListener('chunk_light', saveChunkLight)
          }
          finished = true
        }
        try {
          if (forcedFileNames !== undefined) {
            await ChunkDumper.saveChunkFiles(forcedFileNames.chunkFile, forcedFileNames.metaFile, d)
          } else {
            await ChunkDumper.saveChunkFilesToFolder(folder, d)
          }
          if (finished) {
            resolve()
          }
        } catch (err) {
          this.removeListener('chunk', saveChunk)
          if (this.withLightPackets) {
            this.removeListener('chunk_light', saveChunkLight)
          }
          reject(err)
        }
      }
      this.on('chunk', saveChunk)
      if (this.withLightPackets) {
        saveChunkLight = async d => {
          const { chunkX, chunkZ } = d
          lightsSaved.add(`${chunkX},${chunkZ}`)
          let finished = false
          if (([...lightsSaved].filter(x => chunksSaved.has(x))).length >= count) {
            this.removeListener('chunk', saveChunk)
            this.removeListener('chunk_light', saveChunkLight)
            finished = true
          }
          try {
            if (forcedFileNames !== undefined) {
              await ChunkDumper.saveChunkLightFiles(forcedFileNames.chunkLightFile,
                forcedFileNames.lightMetaFile, d)
            } else {
              await ChunkDumper.saveChunkLightFilesToFolder(folder, d)
            }
            if (finished) {
              resolve()
            }
          } catch (err) {
            this.removeListener('chunk', saveChunk)
            this.removeListener('chunk_light', saveChunkLight)
            reject(err)
          }
        }
        this.on('chunk_light', saveChunkLight)
      }
    })
  }

  async startSavingChunks (folder) {
    try {
      await fs.mkdir(folder)
    } catch (err) {

    }
    this.savingChunk = async d => {
      try {
        await ChunkDumper.saveChunkFilesToFolder(folder, d)
      } catch (err) {
        this.stopSavingChunks()
        throw err
      }
    }
    this.on('chunk', this.savingChunk)
    if (this.withLightPackets) {
      this.savingChunkLight = async d => {
        try {
          await ChunkDumper.saveChunkLightFilesToFolder(folder, d)
        } catch (err) {
          this.stopSavingChunks()
          throw err
        }
      }
      this.on('chunk_light', this.savingChunkLight)
    }
  }

  static async saveChunkFilesToFolder (folder, d) {
    const { x, z } = d
    await ChunkDumper.saveChunkFiles(path.join(folder, 'chunk_' + x + '_' + z + '.dump'),
      path.join(folder, 'chunk_' + x + '_' + z + '.meta'), d)
  }

  static async saveChunkFiles (chunkDataFile, chunkMetaFile, { x, z, groundUp, bitMap, biomes, ignoreOldData, heightmaps, chunkData, skyLightMask, blockLightMask, emptySkyLightMask, emptyBlockLightMask, skyLight, blockLight, trustEdges }) {
    await fs.writeFile(chunkDataFile, chunkData)
    await fs.writeFile(chunkMetaFile, JSON.stringify({
      x, z, groundUp, bitMap, biomes, ignoreOldData, heightmaps, chunkData, skyLightMask, blockLightMask, emptySkyLightMask, emptyBlockLightMask, skyLight, blockLight, trustEdges
    }), 'utf8')
  }

  static async saveChunkLightFilesToFolder (folder, d) {
    const { chunkX, chunkZ } = d
    await ChunkDumper.saveChunkLightFiles(path.join(folder, 'chunk_light_' + chunkX + '_' + chunkZ + '.dump'),
      path.join(folder, 'chunk_light_' + chunkX + '_' + chunkZ + '.meta'), d)
  }

  static async saveChunkLightFiles (chunkLightDataFile, chunkLightMetaFile, {
    chunkX, chunkZ,
    skyLightMask, blockLightMask, emptySkyLightMask, emptyBlockLightMask, skyLight, blockLight, data, trustEdges
  }) {
    if (Buffer.isBuffer(data)) {
      await fs.writeFile(chunkLightDataFile, data)
    } else if (data !== undefined) { // 1.17 doesn't have a data property in their update_light packet
      await fs.writeFile(chunkLightDataFile, JSON.stringify(data))
    }
    await fs.writeFile(chunkLightMetaFile, JSON.stringify({
      chunkX,
      chunkZ,
      skyLightMask,
      blockLightMask,
      emptySkyLightMask,
      emptyBlockLightMask,
      skyLight,
      blockLight,
      trustEdges
    }), 'utf8')
  }

  stopSavingChunks () {
    this.removeListener('chunk', this.savingChunk)
    this.removeListener('chunk_light', this.savingChunkLight)
  }
}

module.exports = ChunkDumper
